community/io/src/main/java/org/neo4j/io/fs/OffsetChannel.java

```diff
@@ -0,0 +1,164 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.io.fs;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileLock;
+
+/**
+ * A wrapper for a channel with an offset. Users must make
+ * sure that the channel is within bounds or exceptions
+ * will occur.
+ */
+public class OffsetChannel implements StoreChannel
+{
+    private final StoreChannel delegate;
+    private final long offset;
+
+    public OffsetChannel( StoreChannel delegate, long offset )
+    {
+        this.delegate = delegate;
+        this.offset = offset;
+    }
+
+    private long offset( long position )
+    {
+        if ( position < 0 )
+        {
+            throw new IllegalArgumentException( "Position must be >= 0." );
+        }
+        return Math.addExact( position, offset );
+    }
+
+    @Override
+    public FileLock tryLock() throws IOException
+    {
+        return delegate.tryLock();
+    }
+
+    @Override
+    public int write( ByteBuffer src, long position ) throws IOException
+    {
+        return delegate.write( src, offset( position ) );
+    }
+
+    @Override
+    public void writeAll( ByteBuffer src, long position ) throws IOException
+    {
+        delegate.writeAll( src, offset( position ) );
+    }
+
+    @Override
+    public void writeAll( ByteBuffer src ) throws IOException
+    {
+        delegate.writeAll( src );
+    }
+
+    @Override
+    public int read( ByteBuffer dst, long position ) throws IOException
+    {
+        return delegate.read( dst, offset( position ) );
+    }
+
+    @Override
+    public void force( boolean metaData ) throws IOException
+    {
+        delegate.force( metaData );
+    }
+
+    @Override
+    public int read( ByteBuffer dst ) throws IOException
+    {
+        return delegate.read( dst );
+    }
+
+    @Override
+    public int write( ByteBuffer src ) throws IOException
+    {
+        return delegate.write( src );
+    }
+
+    @Override
+    public long position() throws IOException
+    {
+        return delegate.position() - offset;
+    }
+
+    @Override
+    public StoreChannel position( long newPosition ) throws IOException
+    {
+        return delegate.position( offset( newPosition ) );
+    }
+
+    @Override
+    public long size() throws IOException
+    {
+        return delegate.size() - offset;
+    }
+
+    @Override
+    public StoreChannel truncate( long size ) throws IOException
+    {
+        return delegate.truncate( offset( size ) );
+    }
+
+    @Override
+    public void flush() throws IOException
+    {
+        delegate.flush();
+    }
+
+    @Override
+    public long write( ByteBuffer[] srcs, int offset, int length ) throws IOException
+    {
+        return delegate.write( srcs, offset, length );
```

Comment:
This offset is an offset into the buffer. There is no need to offset the position since this method does not have a position parameter. Whatever the position was from before will be used.

Reply:
Javadoc states that this channel is a channel with offset, this method and all other that performs operations on ByteBuffer[] are ignoring those which can be confusing

Reply:
I think it is very possible that it is slightly confusing from implementation, but not from the interface though. It should be clear from the interface that the write methods that do not get position as a parameter use the current position. The current position is properly offset by any methods that adjust the position or get the position as a secondary parameter.

This class was just an attempt to improve upon having to think about offsets within the FreeIdKeeper implementation. It was unnecessary to have that extra complexity in there. I don't particularly like the OffsetChannel as such, but it is better than the alternative.